pkg/tools/opentelemetry/opentelemetry.go

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package opentelemetry provides a set of utilities for assisting with telemetry data
+package opentelemetry
+
+import (
+	"context"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/otlp"
+	"go.opentelemetry.io/otel/sdk/metric/controller/push"
+	"go.opentelemetry.io/otel/sdk/metric/processor/basic"
+	"go.opentelemetry.io/otel/sdk/metric/selector/simple"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/semconv"
+	"google.golang.org/grpc"
+	"os"
+	"strconv"
+	"time"
+)
+
+const (
+	// DefaultAgentAddr - opentelemetry agent default address
+	DefaultAgentAddr     = "0.0.0.0:55680"
+
+	opentelemetryEnv     = "TELEMETRY_ENABLED"
+	opentelemetryDefault = true
+
+	defaultPushPeriod    = time.Second
+)
+
+// IsOpentelemetryEnabled returns true if opentelemetry enabled
+func IsOpentelemetryEnabled() bool {
+	val, err := readEnvBool(opentelemetryEnv, opentelemetryDefault)
+	if err == nil {
+		return val
+	}
+	return opentelemetryDefault
+}
+
+func readEnvBool(env string, value bool) (bool, error) {
+	str := os.Getenv(env)
+	if str == "" {
+		return value, nil
+	}
+
+	return strconv.ParseBool(str)
+}
+
+// Init - creates opentelemetry trace and metrics provider
+func Init(ctx context.Context, otelAgentAddr, service string, dialOptions ...grpc.DialOption) (func(), error) {
+	if !IsOpentelemetryEnabled() {
+		return nil, nil
+	}
+
+	exp, err := otlp.NewExporter(
+		ctx,
+		otlp.WithInsecure(),
+		otlp.WithAddress(otelAgentAddr),
+		otlp.WithGRPCDialOption(dialOptions...),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := resource.New(ctx,
+		resource.WithAttributes(
+			// the service name used to display traces in backends
+			semconv.ServiceNameKey.String(service),
+		),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	/* Create tracer provider */
+	bsp := sdktrace.NewBatchSpanProcessor(exp)
+	tracerProvider := sdktrace.NewTracerProvider(
+		sdktrace.WithConfig(sdktrace.Config{DefaultSampler: sdktrace.AlwaysSample()}),
+		sdktrace.WithResource(res),
+		sdktrace.WithSpanProcessor(bsp),
+	)
+	otel.SetTracerProvider(tracerProvider)
+
+	/* Create meter provider*/
+	pusher := push.New(
+		basic.New(
+			simple.NewWithExactDistribution(),
+			exp,
+		),
+		exp,
+		push.WithPeriod(defaultPushPeriod),
```

Comment:
Check if we could send by Request when Request is executed.

Reply:
It seems not.
pusher sends data using tick() only.